File: backend/app/api/comments.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy import select, func, desc, asc
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import joinedload, selectinload
from typing import List, Optional
from datetime import datetime, timezone

from app.database import get_db
from app.models.user import User
from app.models.video import Video
from app.models.comment import Comment
from app.schemas.comment import (
    CommentCreate,
    CommentUpdate,
    CommentResponse,
    CommentListResponse,
    CommentMarker,
)
from app.api.deps import get_current_user

router = APIRouter()


async def get_video_or_404(video_id: str, db: AsyncSession) -> Video:
    """Helper to get video or raise 404."""
    result = await db.execute(select(Video).where(Video.id == video_id))
    video = result.scalar_one_or_none()
    if not video:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Video not found: {video_id}",
        )
    return video


async def get_comment_or_404(comment_id: str, db: AsyncSession) -> Comment:
    """Helper to get comment or raise 404."""
    result = await db.execute(
        select(Comment)
        .options(
            joinedload(Comment.author),
            joinedload(Comment.video),
            selectinload(Comment.replies),
        )
        .where(Comment.id == comment_id)
    )
    comment = result.scalar_one_or_none()
    if not comment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Comment not found: {comment_id}",
        )
    return comment


@router.get("/videos/{video_id}/comments", response_model=CommentListResponse)
async def get_video_comments(
    video_id: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    sort: str = Query("newest", regex="^(newest|oldest|timestamp)$"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get top-level comments for a video with their replies.
    """
    video = await get_video_or_404(video_id, db)

    # Base query for top-level comments
    query = (
        select(Comment)
        .options(
            joinedload(Comment.author),
            selectinload(Comment.replies).joinedload(Comment.author),
        )
        .where(Comment.video_id == video_id)
        .where(Comment.parent_id == None)
    )

    # Apply sorting
    if sort == "newest":
        query = query.order_by(desc(Comment.created_at))
    elif sort == "oldest":
        query = query.order_by(asc(Comment.created_at))
    elif sort == "timestamp":
        # Sort by timestamp (nulls last)
        query = query.order_by(
            Comment.timestamp_seconds.asc().nullslast(), desc(Comment.created_at)
        )

    # Get total count
    count_query = (
        select(func.count(Comment.id))
        .where(Comment.video_id == video_id)
        .where(Comment.parent_id == None)
    )
    total_result = await db.execute(count_query)
    total = total_result.scalar_one()

    # Get paginated results
    result = await db.execute(query.offset(skip).limit(limit))
    comments_orm = result.scalars().all()

    # Build responses
    comments = [
        CommentResponse.from_orm_with_extras(
            c,
            current_user_id=current_user.id,
            video_owner_id=video.uploaded_by,
            is_admin=current_user.role == "admin",
            include_replies=True,
        )
        for c in comments_orm
    ]

    return CommentListResponse(
        comments=comments, total=total, has_more=total > (skip + limit)
    )


@router.post(
    "/videos/{video_id}/comments",
    response_model=CommentResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_comment(
    video_id: str,
    comment_data: CommentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Create a new comment or reply.
    """
    video = await get_video_or_404(video_id, db)

    # If replying, validate parent comment
    if comment_data.parent_id:
        parent = await get_comment_or_404(comment_data.parent_id, db)
        if parent.video_id != video_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Parent comment must belong to the same video",
            )
        if parent.parent_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Nested replies are not supported",
            )

    comment = Comment(
        video_id=video_id,
        user_id=current_user.id,
        content=comment_data.content,
        timestamp_seconds=comment_data.timestamp_seconds,
        parent_id=comment_data.parent_id,
    )

    db.add(comment)
    await db.commit()
    await db.refresh(comment)

    # Reload with author for response
    comment = await get_comment_or_404(comment.id, db)

    return CommentResponse.from_orm_with_extras(
        comment,
        current_user_id=current_user.id,
        video_owner_id=video.uploaded_by,
        is_admin=current_user.role == "admin",
    )


@router.patch("/comments/{comment_id}", response_model=CommentResponse)
async def update_comment(
    comment_id: str,
    comment_data: CommentUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Update a comment. Only the author can update, and only within 24 hours.
    """
    comment = await get_comment_or_404(comment_id, db)

    # Validate author
    if comment.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only edit your own comments",
        )

    # Validate timeframe (24 hours)
    now = datetime.now(timezone.utc)
    created_at = comment.created_at
    if created_at.tzinfo is None:
        created_at = created_at.replace(tzinfo=timezone.utc)

    if (now - created_at).total_seconds() > 86400:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Comment editing window (24h) has expired",
        )

    comment.content = comment_data.content
    await db.commit()
    await db.refresh(comment)

    return CommentResponse.from_orm_with_extras(
        comment,
        current_user_id=current_user.id,
        video_owner_id=comment.video.uploaded_by,
        is_admin=current_user.role == "admin",
    )


@router.delete("/comments/{comment_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_comment(
    comment_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Delete a comment. Author, video owner, or admin can delete.
    """
    comment = await get_comment_or_404(comment_id, db)

    # Validate permissions
    is_author = comment.user_id == current_user.id
    is_video_owner = comment.video.uploaded_by == current_user.id
    is_admin = current_user.role == "admin"

    if not (is_author or is_video_owner or is_admin):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have permission to delete this comment",
        )

    await db.delete(comment)
    await db.commit()

    return None


@router.get("/videos/{video_id}/comment-markers", response_model=List[CommentMarker])
async def get_comment_markers(
    video_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get aggregated comment timestamps for video progress bar.
    """
    await get_video_or_404(video_id, db)

    query = (
        select(Comment.timestamp_seconds, func.count(Comment.id).label("count"))
        .where(Comment.video_id == video_id)
        .where(Comment.timestamp_seconds != None)
        .group_by(Comment.timestamp_seconds)
        .order_by(asc(Comment.timestamp_seconds))
    )

    result = await db.execute(query)
    rows = result.all()

    return [CommentMarker(seconds=row[0], count=row[1]) for row in rows]

File: backend/internal/api/handlers/comments.go
package handlers

import (
	"encoding/json"
	"errors"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	"github.com/clipset/clipset-go/internal/api/middleware"
	"github.com/clipset/clipset-go/internal/api/response"
	"github.com/clipset/clipset-go/internal/config"
	"github.com/clipset/clipset-go/internal/db"
	"github.com/clipset/clipset-go/internal/db/sqlc"
)

// Comment edit window duration
const commentEditWindowHours = 24

// Comment edit threshold in seconds (edits within this window don't count as "edited")
const commentEditThresholdSeconds = 60

// CommentsHandler handles comment management endpoints
type CommentsHandler struct {
	db     *db.DB
	config *config.Config
}

// NewCommentsHandler creates a new comments handler
func NewCommentsHandler(database *db.DB, cfg *config.Config) *CommentsHandler {
	return &CommentsHandler{
		db:     database,
		config: cfg,
	}
}

// --- Response Types ---

// CommentResponse represents a comment with computed fields
type CommentResponse struct {
	ID               string            `json:"id"`
	VideoID          string            `json:"video_id"`
	Content          string            `json:"content"`
	TimestampSeconds *int32            `json:"timestamp_seconds"`
	ParentID         *string           `json:"parent_id"`
	UserID           string            `json:"user_id"`
	AuthorUsername   string            `json:"author_username"`
	AuthorAvatarURL  *string           `json:"author_avatar_url"`
	CreatedAt        time.Time         `json:"created_at"`
	UpdatedAt        time.Time         `json:"updated_at"`
	IsEdited         bool              `json:"is_edited"`
	CanEdit          bool              `json:"can_edit"`
	CanDelete        bool              `json:"can_delete"`
	ReplyCount       int64             `json:"reply_count"`
	Replies          []CommentResponse `json:"replies"`
}

// CommentListResponse represents a paginated list of comments
type CommentListResponse struct {
	Comments []CommentResponse `json:"comments"`
	Total    int64             `json:"total"`
	HasMore  bool              `json:"has_more"`
}

// CommentMarker represents a timestamp marker for the video timeline
type CommentMarker struct {
	Seconds int32 `json:"seconds"`
	Count   int64 `json:"count"`
}

// --- Request Types ---

// CommentCreateRequest represents the create comment request
type CommentCreateRequest struct {
	Content          string  `json:"content"`
	TimestampSeconds *int32  `json:"timestamp_seconds"`
	ParentID         *string `json:"parent_id"`
}

// CommentUpdateRequest represents the update comment request
type CommentUpdateRequest struct {
	Content string `json:"content"`
}

// --- Helper Functions ---

// Note: buildAvatarURL is defined in users.go (same package)

// isEdited checks if a comment has been edited (updated > 60s after creation)
func isEdited(createdAt, updatedAt time.Time) bool {
	return updatedAt.Sub(createdAt).Seconds() > commentEditThresholdSeconds
}

// canEditComment checks if user can edit the comment (author only, within 24h)
func canEditComment(commentUserID, currentUserID uuid.UUID, createdAt time.Time) bool {
	if commentUserID != currentUserID {
		return false
	}
	return time.Since(createdAt) < time.Duration(commentEditWindowHours)*time.Hour
}

// canDeleteComment checks if user can delete the comment (author, video owner, or admin)
func canDeleteComment(commentUserID, videoOwnerID, currentUserID uuid.UUID, isAdmin bool) bool {
	if isAdmin {
		return true
	}
	if commentUserID == currentUserID {
		return true
	}
	if videoOwnerID == currentUserID {
		return true
	}
	return false
}

// pgUUIDToString converts a pgtype.UUID to *string
func pgUUIDToString(u pgtype.UUID) *string {
	if !u.Valid {
		return nil
	}
	s := uuid.UUID(u.Bytes).String()
	return &s
}

// buildCommentResponseFromListRow builds a CommentResponse from a ListCommentsByVideoRow
func buildCommentResponseFromListRow(
	row sqlc.ListCommentsByVideoRow,
	currentUserID, videoOwnerID uuid.UUID,
	isAdmin bool,
	replies []CommentResponse,
) CommentResponse {
	return CommentResponse{
		ID:               row.ID.String(),
		VideoID:          row.VideoID.String(),
		Content:          row.Content,
		TimestampSeconds: row.TimestampSeconds,
		ParentID:         pgUUIDToString(row.ParentID),
		UserID:           row.UserID.String(),
		AuthorUsername:   row.AuthorUsername,
		AuthorAvatarURL:  buildAvatarURL(row.AuthorAvatar),
		CreatedAt:        row.CreatedAt,
		UpdatedAt:        row.UpdatedAt,
		IsEdited:         isEdited(row.CreatedAt, row.UpdatedAt),
		CanEdit:          canEditComment(row.UserID, currentUserID, row.CreatedAt),
		CanDelete:        canDeleteComment(row.UserID, videoOwnerID, currentUserID, isAdmin),
		ReplyCount:       row.ReplyCount,
		Replies:          replies,
	}
}

// buildCommentResponseFromReplyRow builds a CommentResponse from a ListRepliesByCommentRow
func buildCommentResponseFromReplyRow(
	row sqlc.ListRepliesByCommentRow,
	currentUserID, videoOwnerID uuid.UUID,
	isAdmin bool,
) CommentResponse {
	return CommentResponse{
		ID:               row.ID.String(),
		VideoID:          row.VideoID.String(),
		Content:          row.Content,
		TimestampSeconds: row.TimestampSeconds,
		ParentID:         pgUUIDToString(row.ParentID),
		UserID:           row.UserID.String(),
		AuthorUsername:   row.AuthorUsername,
		AuthorAvatarURL:  buildAvatarURL(row.AuthorAvatar),
		CreatedAt:        row.CreatedAt,
		UpdatedAt:        row.UpdatedAt,
		IsEdited:         isEdited(row.CreatedAt, row.UpdatedAt),
		CanEdit:          canEditComment(row.UserID, currentUserID, row.CreatedAt),
		CanDelete:        canDeleteComment(row.UserID, videoOwnerID, currentUserID, isAdmin),
		ReplyCount:       0,
		Replies:          nil,
	}
}

// buildCommentResponseFromWithAuthorRow builds a CommentResponse from GetCommentWithAuthorRow
func buildCommentResponseFromWithAuthorRow(
	row sqlc.GetCommentWithAuthorRow,
	currentUserID uuid.UUID,
	isAdmin bool,
	replyCount int64,
) CommentResponse {
	return CommentResponse{
		ID:               row.ID.String(),
		VideoID:          row.VideoID.String(),
		Content:          row.Content,
		TimestampSeconds: row.TimestampSeconds,
		ParentID:         pgUUIDToString(row.ParentID),
		UserID:           row.UserID.String(),
		AuthorUsername:   row.AuthorUsername,
		AuthorAvatarURL:  buildAvatarURL(row.AuthorAvatar),
		CreatedAt:        row.CreatedAt,
		UpdatedAt:        row.UpdatedAt,
		IsEdited:         isEdited(row.CreatedAt, row.UpdatedAt),
		CanEdit:          canEditComment(row.UserID, currentUserID, row.CreatedAt),
		CanDelete:        canDeleteComment(row.UserID, row.VideoOwnerID, currentUserID, isAdmin),
		ReplyCount:       replyCount,
		Replies:          nil,
	}
}

// --- Handlers ---

// ListByVideo handles GET /api/videos/{video_id}/comments
func (h *CommentsHandler) ListByVideo(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Parse video_id
	videoIDStr := r.PathValue("video_id")
	if videoIDStr == "" {
		response.BadRequest(w, "Video ID is required")
		return
	}

	videoID, err := uuid.Parse(videoIDStr)
	if err != nil {
		response.BadRequest(w, "Invalid video ID format")
		return
	}

	// Get current user
	currentUserID, ok := middleware.GetUserID(ctx)
	if !ok {
		response.Unauthorized(w, "Not authenticated")
		return
	}
	isAdmin := middleware.IsAdmin(ctx)

	// Verify video exists and get owner
	video, err := h.db.Queries.GetVideoByID(ctx, videoID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			response.NotFound(w, "Video not found")
			return
		}
		log.Printf("Error getting video: %v", err)
		response.InternalServerError(w, "Failed to get video")
		return
	}
	videoOwnerID := video.UploadedBy

	// Parse pagination parameters
	skip := 0
	limit := 50
	sort := "newest"

	if skipStr := r.URL.Query().Get("skip"); skipStr != "" {
		if s, err := strconv.Atoi(skipStr); err == nil && s >= 0 {
			skip = s
		}
	}

	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
		if l, err := strconv.Atoi(limitStr); err == nil && l >= 1 && l <= 100 {
			limit = l
		}
	}

	if sortParam := r.URL.Query().Get("sort"); sortParam != "" {
		switch sortParam {
		case "newest", "oldest", "timestamp":
			sort = sortParam
		default:
			response.BadRequest(w, "Invalid sort parameter. Must be 'newest', 'oldest', or 'timestamp'")
			return
		}
	}

	// Get top-level comments
	comments, err := h.db.Queries.ListCommentsByVideo(ctx, sqlc.ListCommentsByVideoParams{
		VideoID: videoID,
		Column2: sort,
		Limit:   int32(limit),
		Offset:  int32(skip),
	})
	if err != nil {
		log.Printf("Error listing comments: %v", err)
		response.InternalServerError(w, "Failed to get comments")
		return
	}

	// Get total count
	total, err := h.db.Queries.CountCommentsByVideo(ctx, videoID)
	if err != nil {
		log.Printf("Error counting comments: %v", err)
		response.InternalServerError(w, "Failed to count comments")
		return
	}

	// Fetch all replies for all top-level comments eagerly
	// Build a map of parent_id -> replies
	repliesMap := make(map[uuid.UUID][]CommentResponse)

	for _, comment := range comments {
		replies, err := h.db.Queries.ListRepliesByComment(ctx, pgtype.UUID{
			Bytes: comment.ID,
			Valid: true,
		})
		if err != nil {
			log.Printf("Error fetching replies for comment %s: %v", comment.ID, err)
			continue
		}

		replyResponses := make([]CommentResponse, len(replies))
		for i, reply := range replies {
			replyResponses[i] = buildCommentResponseFromReplyRow(reply, currentUserID, videoOwnerID, isAdmin)
		}
		repliesMap[comment.ID] = replyResponses
	}

	// Build response
	commentResponses := make([]CommentResponse, len(comments))
	for i, comment := range comments {
		replies := repliesMap[comment.ID]
		if replies == nil {
			replies = []CommentResponse{}
		}
		commentResponses[i] = buildCommentResponseFromListRow(comment, currentUserID, videoOwnerID, isAdmin, replies)
	}

	hasMore := total > int64(skip+limit)

	response.OK(w, CommentListResponse{
		Comments: commentResponses,
		Total:    total,
		HasMore:  hasMore,
	})
}

// Create handles POST /api/videos/{video_id}/comments
func (h *CommentsHandler) Create(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Parse video_id
	videoIDStr := r.PathValue("video_id")
	if videoIDStr == "" {
		response.BadRequest(w, "Video ID is required")
		return
	}

	videoID, err := uuid.Parse(videoIDStr)
	if err != nil {
		response.BadRequest(w, "Invalid video ID format")
		return
	}

	// Get current user
	currentUserID, ok := middleware.GetUserID(ctx)
	if !ok {
		response.Unauthorized(w, "Not authenticated")
		return
	}
	currentUsername, _ := middleware.GetUsername(ctx)
	isAdmin := middleware.IsAdmin(ctx)

	// Verify video exists and get owner
	video, err := h.db.Queries.GetVideoByID(ctx, videoID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			response.NotFound(w, "Video not found")
			return
		}
		log.Printf("Error getting video: %v", err)
		response.InternalServerError(w, "Failed to get video")
		return
	}
	videoOwnerID := video.UploadedBy

	// Parse request body
	var req CommentCreateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		response.BadRequest(w, "Invalid request body")
		return
	}

	// Validate content
	content := strings.TrimSpace(req.Content)
	if content == "" {
		response.BadRequest(w, "Content is required")
		return
	}
	if len(content) > 2000 {
		response.BadRequest(w, "Content must be 2000 characters or less")
		return
	}

	// Validate timestamp_seconds
	if req.TimestampSeconds != nil && *req.TimestampSeconds < 0 {
		response.BadRequest(w, "Timestamp must be >= 0")
		return
	}

	// Handle parent_id for replies
	var parentID pgtype.UUID
	if req.ParentID != nil && *req.ParentID != "" {
		parsedParentID, err := uuid.Parse(*req.ParentID)
		if err != nil {
			response.BadRequest(w, "Invalid parent ID format")
			return
		}

		// Fetch parent comment
		parentComment, err := h.db.Queries.GetCommentByID(ctx, parsedParentID)
		if err != nil {
			if errors.Is(err, pgx.ErrNoRows) {
				response.NotFound(w, "Parent comment not found")
				return
			}
			log.Printf("Error getting parent comment: %v", err)
			response.InternalServerError(w, "Failed to validate parent comment")
			return
		}

		// Verify parent belongs to same video
		if parentComment.VideoID != videoID {
			response.BadRequest(w, "Parent comment belongs to a different video")
			return
		}

		// Verify parent is not a reply (single-level nesting only)
		if parentComment.ParentID.Valid {
			response.BadRequest(w, "Nested replies are not supported")
			return
		}

		parentID = pgtype.UUID{Bytes: parsedParentID, Valid: true}
	}

	// Create comment
	comment, err := h.db.Queries.CreateComment(ctx, sqlc.CreateCommentParams{
		VideoID:          videoID,
		UserID:           currentUserID,
		Content:          content,
		TimestampSeconds: req.TimestampSeconds,
		ParentID:         parentID,
	})
	if err != nil {
		log.Printf("Error creating comment: %v", err)
		response.InternalServerError(w, "Failed to create comment")
		return
	}

	// Get user avatar for response
	user, err := h.db.Queries.GetUserByID(ctx, currentUserID)
	if err != nil {
		log.Printf("Warning: couldn't get user info for response: %v", err)
	}

	var avatarURL *string
	if user.AvatarFilename != nil {
		avatarURL = buildAvatarURL(user.AvatarFilename)
	}

	log.Printf("Created comment %s on video %s by user %s", comment.ID, videoID, currentUserID)

	response.Created(w, CommentResponse{
		ID:               comment.ID.String(),
		VideoID:          comment.VideoID.String(),
		Content:          comment.Content,
		TimestampSeconds: comment.TimestampSeconds,
		ParentID:         pgUUIDToString(comment.ParentID),
		UserID:           comment.UserID.String(),
		AuthorUsername:   currentUsername,
		AuthorAvatarURL:  avatarURL,
		CreatedAt:        comment.CreatedAt,
		UpdatedAt:        comment.UpdatedAt,
		IsEdited:         false,
		CanEdit:          canEditComment(comment.UserID, currentUserID, comment.CreatedAt),
		CanDelete:        canDeleteComment(comment.UserID, videoOwnerID, currentUserID, isAdmin),
		ReplyCount:       0,
		Replies:          nil,
	})
}

// Update handles PATCH /api/comments/{comment_id}
func (h *CommentsHandler) Update(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Parse comment_id
	commentIDStr := r.PathValue("comment_id")
	if commentIDStr == "" {
		response.BadRequest(w, "Comment ID is required")
		return
	}

	commentID, err := uuid.Parse(commentIDStr)
	if err != nil {
		response.BadRequest(w, "Invalid comment ID format")
		return
	}

	// Get current user
	currentUserID, ok := middleware.GetUserID(ctx)
	if !ok {
		response.Unauthorized(w, "Not authenticated")
		return
	}
	isAdmin := middleware.IsAdmin(ctx)

	// Get comment with author and video owner
	comment, err := h.db.Queries.GetCommentWithAuthor(ctx, commentID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			response.NotFound(w, "Comment not found")
			return
		}
		log.Printf("Error getting comment: %v", err)
		response.InternalServerError(w, "Failed to get comment")
		return
	}

	// Check permission: author only can edit
	if comment.UserID != currentUserID {
		response.Forbidden(w, "You can only edit your own comments")
		return
	}

	// Check time window: within 24 hours
	if time.Since(comment.CreatedAt) >= time.Duration(commentEditWindowHours)*time.Hour {
		response.Forbidden(w, "Comment editing window (24h) has expired")
		return
	}

	// Parse request body
	var req CommentUpdateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		response.BadRequest(w, "Invalid request body")
		return
	}

	// Validate content
	content := strings.TrimSpace(req.Content)
	if content == "" {
		response.BadRequest(w, "Content is required")
		return
	}
	if len(content) > 2000 {
		response.BadRequest(w, "Content must be 2000 characters or less")
		return
	}

	// Update comment
	updatedComment, err := h.db.Queries.UpdateComment(ctx, sqlc.UpdateCommentParams{
		ID:      commentID,
		Content: content,
	})
	if err != nil {
		log.Printf("Error updating comment: %v", err)
		response.InternalServerError(w, "Failed to update comment")
		return
	}

	// Get reply count if this is a top-level comment
	var replyCount int64
	if !comment.ParentID.Valid {
		replyCount, _ = h.db.Queries.CountCommentsByVideo(ctx, comment.VideoID)
	}

	log.Printf("Updated comment %s by user %s", commentID, currentUserID)

	response.OK(w, CommentResponse{
		ID:               updatedComment.ID.String(),
		VideoID:          updatedComment.VideoID.String(),
		Content:          updatedComment.Content,
		TimestampSeconds: updatedComment.TimestampSeconds,
		ParentID:         pgUUIDToString(updatedComment.ParentID),
		UserID:           updatedComment.UserID.String(),
		AuthorUsername:   comment.AuthorUsername,
		AuthorAvatarURL:  buildAvatarURL(comment.AuthorAvatar),
		CreatedAt:        updatedComment.CreatedAt,
		UpdatedAt:        updatedComment.UpdatedAt,
		IsEdited:         isEdited(updatedComment.CreatedAt, updatedComment.UpdatedAt),
		CanEdit:          canEditComment(updatedComment.UserID, currentUserID, updatedComment.CreatedAt),
		CanDelete:        canDeleteComment(updatedComment.UserID, comment.VideoOwnerID, currentUserID, isAdmin),
		ReplyCount:       replyCount,
		Replies:          nil,
	})
}

// Delete handles DELETE /api/comments/{comment_id}
func (h *CommentsHandler) Delete(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Parse comment_id
	commentIDStr := r.PathValue("comment_id")
	if commentIDStr == "" {
		response.BadRequest(w, "Comment ID is required")
		return
	}

	commentID, err := uuid.Parse(commentIDStr)
	if err != nil {
		response.BadRequest(w, "Invalid comment ID format")
		return
	}

	// Get current user
	currentUserID, ok := middleware.GetUserID(ctx)
	if !ok {
		response.Unauthorized(w, "Not authenticated")
		return
	}
	isAdmin := middleware.IsAdmin(ctx)

	// Get comment with author and video owner
	comment, err := h.db.Queries.GetCommentWithAuthor(ctx, commentID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			response.NotFound(w, "Comment not found")
			return
		}
		log.Printf("Error getting comment: %v", err)
		response.InternalServerError(w, "Failed to get comment")
		return
	}

	// Check permission: author, video owner, or admin
	if !canDeleteComment(comment.UserID, comment.VideoOwnerID, currentUserID, isAdmin) {
		response.Forbidden(w, "You don't have permission to delete this comment")
		return
	}

	// Delete comment (cascade handles replies)
	if err := h.db.Queries.DeleteComment(ctx, commentID); err != nil {
		log.Printf("Error deleting comment: %v", err)
		response.InternalServerError(w, "Failed to delete comment")
		return
	}

	log.Printf("Deleted comment %s by user %s", commentID, currentUserID)

	response.NoContent(w)
}

// GetMarkers handles GET /api/videos/{video_id}/comment-markers
func (h *CommentsHandler) GetMarkers(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Parse video_id
	videoIDStr := r.PathValue("video_id")
	if videoIDStr == "" {
		response.BadRequest(w, "Video ID is required")
		return
	}

	videoID, err := uuid.Parse(videoIDStr)
	if err != nil {
		response.BadRequest(w, "Invalid video ID format")
		return
	}

	// Verify user is authenticated
	_, ok := middleware.GetUserID(ctx)
	if !ok {
		response.Unauthorized(w, "Not authenticated")
		return
	}

	// Verify video exists
	_, err = h.db.Queries.GetVideoByID(ctx, videoID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			response.NotFound(w, "Video not found")
			return
		}
		log.Printf("Error getting video: %v", err)
		response.InternalServerError(w, "Failed to get video")
		return
	}

	// Get markers
	markers, err := h.db.Queries.GetCommentMarkers(ctx, videoID)
	if err != nil {
		log.Printf("Error getting comment markers: %v", err)
		response.InternalServerError(w, "Failed to get comment markers")
		return
	}

	// Build response
	markerResponses := make([]CommentMarker, len(markers))
	for i, m := range markers {
		// timestamp_seconds should never be nil here due to the WHERE clause,
		// but handle it defensively
		seconds := int32(0)
		if m.TimestampSeconds != nil {
			seconds = *m.TimestampSeconds
		}
		markerResponses[i] = CommentMarker{
			Seconds: seconds,
			Count:   m.Count,
		}
	}

	response.OK(w, markerResponses)
}
